File: agents/recruitment-assistant-agent/src/recruitment_assistant_agent/api/v1/endpoints/admin_api_config.py
"""Current-user endpoints for the agent API config center (CONFIG-100)."""

from __future__ import annotations

from datetime import UTC, datetime

from fastapi import APIRouter

from recruitment_assistant_agent.api.dependencies import (
    AgentApiConfigCryptoDep,
    AgentApiConfigRepoDep,
    CurrentUserPublicIdDep,
)
from recruitment_assistant_agent.api.v1.schemas.agent_api_config import (
    AgentApiConfigEnvelope,
    AgentApiConfigListData,
    AgentApiConfigListEnvelope,
    AgentApiConfigUpdate,
    AgentApiConfigView,
)
from recruitment_assistant_agent.core.errors import ConfigNotFoundError

router = APIRouter()

_API_TYPE_DEFAULTS = {
    "deepseek": "DeepSeek LLM（必需）",
    "ocr": "OCR（扫描简历/JD 可选）",
}
_ALLOWED_API_TYPES = frozenset(_API_TYPE_DEFAULTS)


def _to_view(entity) -> AgentApiConfigView:  # type: ignore[no-untyped-def]
    return AgentApiConfigView(
        user_public_id=entity.user_public_id,
        api_type=entity.api_type,
        display_name=entity.display_name,
        base_url=entity.base_url,
        model=entity.model,
        api_key_hint=entity.api_key_hint,
        timeout_seconds=entity.timeout_seconds,
        max_retries=entity.max_retries,
        enabled=entity.enabled,
        extra=entity.extra,
        updated_by=entity.updated_by,
        updated_at=entity.updated_at,
    )


def _empty_view(user_public_id: str, api_type: str) -> AgentApiConfigView:
    return AgentApiConfigView(
        user_public_id=user_public_id,
        api_type=api_type,
        display_name=_API_TYPE_DEFAULTS[api_type],
        base_url=None,
        model=None,
        api_key_hint=None,
        timeout_seconds=None,
        max_retries=None,
        enabled=False,
        extra=None,
        updated_by=user_public_id,
        updated_at=datetime.now(UTC),
    )


@router.get(
    "/me/api-config",
    response_model=AgentApiConfigListEnvelope,
)
async def list_api_configs(
    user_public_id: CurrentUserPublicIdDep,
    repo: AgentApiConfigRepoDep,
) -> AgentApiConfigListEnvelope:
    """List upstream API dependencies configured by the current user."""

    items = await repo.list_for_user(user_public_id)
    views = {_item.api_type: _to_view(_item) for _item in items}
    for api_type in _API_TYPE_DEFAULTS:
        views.setdefault(api_type, _empty_view(user_public_id, api_type))
    return AgentApiConfigListEnvelope(data=AgentApiConfigListData(items=list(views.values())))


@router.get(
    "/me/api-config/{api_type}",
    response_model=AgentApiConfigEnvelope,
)
async def get_api_config(
    api_type: str,
    user_public_id: CurrentUserPublicIdDep,
    repo: AgentApiConfigRepoDep,
) -> AgentApiConfigEnvelope:
    """Fetch one current-user API dependency by api_type."""

    if api_type not in _ALLOWED_API_TYPES:
        raise ConfigNotFoundError(f"unknown api_type: {api_type}")
    entity = await repo.get_by_type(user_public_id, api_type)
    if entity is None:
        raise ConfigNotFoundError(f"api_type not configured: {api_type}")
    return AgentApiConfigEnvelope(data=_to_view(entity))


@router.put(
    "/me/api-config/{api_type}",
    response_model=AgentApiConfigEnvelope,
)
async def update_api_config(
    api_type: str,
    payload: AgentApiConfigUpdate,
    user_public_id: CurrentUserPublicIdDep,
    repo: AgentApiConfigRepoDep,
    crypto: AgentApiConfigCryptoDep,
) -> AgentApiConfigEnvelope:
    """Create or update a current-user API dependency.

    `api_key` is encrypted at write time; only the masked hint is returned.
    An empty `api_key` keeps the existing encrypted value unchanged.
    """

    if api_type not in _ALLOWED_API_TYPES:
        raise ConfigNotFoundError(f"unknown api_type: {api_type}")

    existing = await repo.get_by_type(user_public_id, api_type)
    api_key_encrypted = existing.api_key_encrypted if existing is not None else None
    api_key_hint = existing.api_key_hint if existing is not None else None

    if payload.api_key:
        api_key_encrypted = crypto.encrypt(payload.api_key)
        api_key_hint = crypto.hint(payload.api_key)

    entity = await repo.upsert(
        user_public_id=user_public_id,
        api_type=api_type,
        display_name=payload.display_name,
        base_url=payload.base_url,
        model=payload.model,
        api_key_encrypted=api_key_encrypted,
        api_key_hint=api_key_hint,
        timeout_seconds=payload.timeout_seconds,
        max_retries=payload.max_retries,
        enabled=payload.enabled,
        extra=payload.extra,
        updated_by=user_public_id,
    )
    return AgentApiConfigEnvelope(data=_to_view(entity))

File: agent-suite-web/src/api/admin-api-config.ts
import type { AxiosInstance } from 'axios';

import { dataApi, legalApi, recruitmentApi } from '@lib/config';
import type { SuiteAxiosRequestConfig } from '@lib/http-client';
import type {
  AgentApiConfigEnvelope,
  AgentApiConfigListEnvelope,
  AgentApiConfigUpdate,
  AgentApiConfigView,
  AgentName,
} from '@/types/admin-api-config';

const clientsByAgent: Record<AgentName, AxiosInstance> = {
  legal: legalApi,
  recruitment: recruitmentApi,
  data: dataApi,
};

const defaultApiTypesByAgent: Record<AgentName, { api_type: string; display_name: string }[]> = {
  legal: [
    { api_type: 'deepseek', display_name: 'DeepSeek LLM（必需）' },
    { api_type: 'embedding', display_name: 'Embedding 服务（法律 RAG）' },
    { api_type: 'vector_db', display_name: '向量库（法律 RAG）' },
    { api_type: 'reranker', display_name: 'Reranker（法律 RAG）' },
    { api_type: 'ocr', display_name: 'OCR（扫描材料可选）' },
  ],
  recruitment: [
    { api_type: 'deepseek', display_name: 'DeepSeek LLM（必需）' },
    { api_type: 'ocr', display_name: 'OCR（扫描简历/JD 可选）' },
  ],
  data: [
    { api_type: 'deepseek', display_name: 'DeepSeek LLM（必需）' },
    { api_type: 'mcp', display_name: 'MCP 查询服务（真查询）' },
    { api_type: 'feishu', display_name: '飞书（消息入口可选）' },
  ],
};

export function buildDefaultAgentApiConfigs(
  agent: AgentName,
  userPublicId = '',
): AgentApiConfigView[] {
  return defaultApiTypesByAgent[agent].map((item) => ({
    user_public_id: userPublicId,
    api_type: item.api_type,
    display_name: item.display_name,
    base_url: null,
    model: null,
    api_key_hint: null,
    timeout_seconds: null,
    max_retries: null,
    enabled: false,
    extra: null,
    updated_by: userPublicId,
    updated_at: new Date(0).toISOString(),
  }));
}

export async function listAgentApiConfigs(agent: AgentName): Promise<AgentApiConfigView[]> {
  const client = clientsByAgent[agent];
  const requestConfig: SuiteAxiosRequestConfig = { suppressGlobalError: true };
  const response = await client.get<AgentApiConfigListEnvelope>('/me/api-config', requestConfig);
  const items = response.data.data.items;
  const userPublicId = items[0]?.user_public_id ?? '';
  const merged = new Map(
    buildDefaultAgentApiConfigs(agent, userPublicId).map((item) => [item.api_type, item]),
  );
  for (const item of items) {
    merged.set(item.api_type, item);
  }
  return Array.from(merged.values());
}

export async function updateAgentApiConfig(
  agent: AgentName,
  apiType: string,
  payload: AgentApiConfigUpdate,
): Promise<AgentApiConfigView> {
  const client = clientsByAgent[agent];
  const response = await client.put<AgentApiConfigEnvelope>(`/me/api-config/${apiType}`, payload);
  return response.data.data;
}
